### PR TITLE
Add new withDockerfile methods to ImageFromDockerfile

### PR DESCRIPTION
**Pull request by aguibert**
These methods are preferable to the existing withDockerfilePath
methods because the docker-java API will honor .dockerignore files
if withDockerfile(File) is used.

### CODE REVIEW

**Comment by aguibert** [comment]
Fixes #1534

File: core/src/main/java/org/testcontainers/images/builder/ImageFromDockerfile.java [hunk 1]
```diff
@@ -156,8 +168,24 @@ public ImageFromDockerfile withBuildArgs(final Map<String, String> args) {
         return this;
     }
 
+    /**
+     * Sets the Dockerfile to be used for this image. 
+     * It is recommended to use {@link #withDockerfile} instead because it honors .dockerignore
+     * files and therefore will be more efficient
+     * @param relativePathFromBuildRoot
+     */
     public ImageFromDockerfile withDockerfilePath(String relativePathFromBuildRoot) {
```
Comment (rnorth) [inline core/src/main/java/org/testcontainers/images/builder/ImageFromDockerfile.java]: I wonder, would it be possible to make the existing `withDockerfilePath` take advantage of `buildImageCmd::withDockerfile` as well? I think it would be nice to change this PR so that:

* Existing users of the `withDockerfilePath` method get the benefits of `.dockerignore`
* We don't need to add (an effectively duplicate) method for `withDockerfile(String pathname)`

File: core/src/main/java/org/testcontainers/images/builder/ImageFromDockerfile.java [hunk 2]
```diff
@@ -128,9 +132,15 @@ public void onNext(BuildResponseItem item) {
...
                     final String destination = entry.getKey();
                     transferable.transferTo(tarArchive, destination);
+                    bytesToDockerDaemon += transferable.getSize();
                 }
                 tarArchive.finish();
             }
+            
+            log.info("Transferred {} KB to Docker daemon", bytesToDockerDaemon % 1024);
+            if (bytesToDockerDaemon > 50 * 1024 * 1024) // warn if >50MB sent to docker daemon
+                log.warn("A large amount of data was sent to the Docker daemon ({}). Consider using a .dockerignore file for better performance.",
```
Comment (rnorth) [inline core/src/main/java/org/testcontainers/images/builder/ImageFromDockerfile.java]: This is a great idea to add 👍 !

File: core/src/main/java/org/testcontainers/images/builder/ImageFromDockerfile.java [hunk 3]
```diff
@@ -128,9 +132,15 @@ public void onNext(BuildResponseItem item) {
...
                     transferable.transferTo(tarArchive, destination);
+                    bytesToDockerDaemon += transferable.getSize();
                 }
                 tarArchive.finish();
             }
+            
+            log.info("Transferred {} KB to Docker daemon", bytesToDockerDaemon % 1024);
+            if (bytesToDockerDaemon > 50 * 1024 * 1024) // warn if >50MB sent to docker daemon
+                log.warn("A large amount of data was sent to the Docker daemon ({}). Consider using a .dockerignore file for better performance.", 
+                        bytesToDockerDaemon % (1024 * 1024));
```
Comment (rnorth) [inline core/src/main/java/org/testcontainers/images/builder/ImageFromDockerfile.java]: As above, shouldn't it be?
```suggestion
                        bytesToDockerDaemon / (1024 * 1024));
```
and add `MB` to the log message so that it's clear.

File: core/src/main/java/org/testcontainers/images/builder/ImageFromDockerfile.java [hunk 4]
```diff
@@ -128,9 +132,15 @@ public void onNext(BuildResponseItem item) {
                     Transferable transferable = entry.getValue();
                     final String destination = entry.getKey();
                     transferable.transferTo(tarArchive, destination);
+                    bytesToDockerDaemon += transferable.getSize();
                 }
                 tarArchive.finish();
             }
+            
+            log.info("Transferred {} KB to Docker daemon", bytesToDockerDaemon % 1024);
```
Comment (rnorth) [inline core/src/main/java/org/testcontainers/images/builder/ImageFromDockerfile.java]: Are you sure about this calculation?

Shouldn't it be:
```suggestion
            log.info("Transferred {} KB to Docker daemon", bytesToDockerDaemon / 1024);
```

Comment (rnorth) [reply core/src/main/java/org/testcontainers/images/builder/ImageFromDockerfile.java]: Also, why not use a consistent unit (either KB or MB) for both log messages?

File: core/src/main/java/org/testcontainers/images/builder/ImageFromDockerfile.java (hunk 1 above)
Comment (aguibert) [reply core/src/main/java/org/testcontainers/images/builder/ImageFromDockerfile.java]: Good suggestion, I had thought of this too and experimented a bit initially. Unfortunately I don't think this is feasible because we'd have to take into account things like setting the docker build path. I'm not sure if I'm missing any more, but I'm not familiar enough with the docker-java API (yet) to confidently make this change with certainty it will not break existing users.

File: core/src/main/java/org/testcontainers/images/builder/ImageFromDockerfile.java (hunk 3 above)
Comment (aguibert) [reply core/src/main/java/org/testcontainers/images/builder/ImageFromDockerfile.java]: yes, good catch!

File: core/src/main/java/org/testcontainers/images/builder/ImageFromDockerfile.java (hunk 4 above)
Comment (aguibert) [reply core/src/main/java/org/testcontainers/images/builder/ImageFromDockerfile.java]: I wanted to use MB for the "woa you transferred a lot of data" message, because at that order of magnitude it will be more easily readable. However, for this log statement there will be a lot of cases where <1MB will be transferred and I didn't want to report "0 MB of data transferred" because that's confusing. Also didn't think it was worth branching the code to do "report in KB if <1MB otherwise report in MB"

File: core/src/main/java/org/testcontainers/images/builder/ImageFromDockerfile.java (hunk 1 above)
Comment (kiview) [reply core/src/main/java/org/testcontainers/images/builder/ImageFromDockerfile.java]: Shoud we deperecate the old method in this case?

File: core/src/main/java/org/testcontainers/images/builder/ImageFromDockerfile.java [hunk 5]
```diff
@@ -156,8 +168,24 @@ public ImageFromDockerfile withBuildArgs(final Map<String, String> args) {
...
         this.dockerFilePath = Optional.of(relativePathFromBuildRoot);
         return this;
     }
+    
+    public ImageFromDockerfile withDockerfile(File dockerfile) {
+        this.dockerfile = Optional.of(dockerfile);
+        return this;
+    }
+    
+    public ImageFromDockerfile withDockerfile(String pathname) {
```
Comment (bsideup) [inline core/src/main/java/org/testcontainers/images/builder/ImageFromDockerfile.java]: let's not have this method, since it is unclear whether it will be using the classpath or FS

File: core/src/main/java/org/testcontainers/images/builder/ImageFromDockerfile.java [hunk 6]
```diff
@@ -63,6 +64,7 @@
     private final Map<String, Transferable> transferables = new HashMap<>();
     private final Map<String, String> buildArgs = new HashMap<>();
     private Optional<String> dockerFilePath = Optional.empty();
+    private Optional<File> dockerfile = Optional.empty();
```
Comment (bsideup) [inline core/src/main/java/org/testcontainers/images/builder/ImageFromDockerfile.java]: can we have Path here please?

File: core/src/main/java/org/testcontainers/images/builder/ImageFromDockerfile.java (hunk 4 above)
Comment (rnorth) [reply core/src/main/java/org/testcontainers/images/builder/ImageFromDockerfile.java]: We happen to have commons-io `FileUtils` on the classpath - we could actually delegate all this formatting to its `byteCountToDisplaySize` method, and not have to worry about which is the most appropriate unit for either message. WDYT?

File: core/src/main/java/org/testcontainers/images/builder/ImageFromDockerfile.java (hunk 1 above)
Comment (rnorth) [reply core/src/main/java/org/testcontainers/images/builder/ImageFromDockerfile.java]: That makes sense @aguibert - I can see how complexity could run away in trying to make it work.

I agree with @kiview re deprecating the existing method

File: core/src/main/java/org/testcontainers/images/builder/ImageFromDockerfile.java (hunk 5 above)
Comment (rnorth) [reply core/src/main/java/org/testcontainers/images/builder/ImageFromDockerfile.java]: Good point. It could also be a dockerfile as a string!

I don't see why we couldn't remove this method.

File: core/src/main/java/org/testcontainers/images/builder/ImageFromDockerfile.java (hunk 6 above)
Comment (rnorth) [reply core/src/main/java/org/testcontainers/images/builder/ImageFromDockerfile.java]: @bsideup do you mean hold an `Optional<Path>` object in the field _and_ make the with-setter accept a `Path`, or just change the field type?

File: core/src/main/java/org/testcontainers/images/builder/ImageFromDockerfile.java [hunk 7]
```diff
@@ -128,9 +132,15 @@ public void onNext(BuildResponseItem item) {
                     Transferable transferable = entry.getValue();
                     final String destination = entry.getKey();
                     transferable.transferTo(tarArchive, destination);
+                    bytesToDockerDaemon += transferable.getSize();
                 }
                 tarArchive.finish();
             }
+            
+            log.info("Transferred {} KB to Docker daemon", bytesToDockerDaemon % 1024);
+            if (bytesToDockerDaemon > 50 * 1024 * 1024) // warn if >50MB sent to docker daemon
```
[the remark this answers is not in the document]
Comment (aguibert) [reply core/src/main/java/org/testcontainers/images/builder/ImageFromDockerfile.java]: I see commons-io `FileUtils` has some constants so I'll use those instead

File: core/src/main/java/org/testcontainers/images/builder/ImageFromDockerfile.java (hunk 5 above)
Comment (aguibert) [reply core/src/main/java/org/testcontainers/images/builder/ImageFromDockerfile.java]: I was thinking it may be convenient so people don't have to do `withDockerfile(new FIle("Dockerfile"))` in the normal case. Since it may cause confusion I'll remove it though

File: core/src/main/java/org/testcontainers/images/builder/ImageFromDockerfile.java (hunk 6 above)
Comment (aguibert) [reply core/src/main/java/org/testcontainers/images/builder/ImageFromDockerfile.java]: I have `File` here because the underlying docker-java API takes a `File`, so I thought it would be best to not do the unnecessary conversion. Is there a particular reason we should be using `Path` over `File`?

**Comment by aguibert** [comment]
@rnorth @kiview @bsideup thanks for the reviews! I've updated my PR which I believe addresses all of the review comments thus far -- please have another look when you get some time

File: core/src/main/java/org/testcontainers/images/builder/ImageFromDockerfile.java (hunk 6 above)
Comment (bsideup) [reply core/src/main/java/org/testcontainers/images/builder/ImageFromDockerfile.java]: `Path` is a bit more forward thinking abstraction, easier to work with and I hope docker-java will adopt `Path` API soon too, so that we can avoid temporary files when it is possible.
I would prefer to have it in Testcontainers, because that's our public API, and to not block us from making changes to it later :)

File: core/src/test/java/org/testcontainers/images/builder/DockerignoreTest.java [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+            fail("Should not be able to build an image with an invalid .dockerignore file");
+        } catch(DockerClientException e) {
+            if (!e.getMessage().contains("Invalid pattern"))
+                throw e;
+        }
+    }
+    
+    @Test
+    public void testValidDockerignore() throws Exception {
+        try (final GenericContainer container = new GenericContainer(new ImageFromDockerfile()
```
Comment (bsideup) [inline core/src/test/java/org/testcontainers/images/builder/DockerignoreTest.java]: nit1: please extract the image definition

nit2: could you please add a line break after `try (` and also before `) {`? I know that we're a bit inconsistent, but usually it helps, especially with long fluent-style chains

File: core/src/test/resources/dockerfile-build-test/should_not_be_ignored.txt [hunk 9]
```diff
@@ -0,0 +1 @@
+this file should be ignored by a .dockerignore file
```
Comment (bsideup) [inline core/src/test/resources/dockerfile-build-test/should_not_be_ignored.txt]: filename says that it should not be ignored

File: core/src/main/java/org/testcontainers/images/builder/ImageFromDockerfile.java [hunk 10]
```diff
@@ -156,8 +170,24 @@ public ImageFromDockerfile withBuildArgs(final Map<String, String> args) {
         return this;
     }
 
+    /**
+     * Sets the Dockerfile to be used for this image. 
+     * @deprecated It is recommended to use {@link #withDockerfile} instead because it honors
```
Comment (bsideup) [inline core/src/main/java/org/testcontainers/images/builder/ImageFromDockerfile.java]: Could you please add `@Deprecated` to the method?

File: core/src/test/java/org/testcontainers/images/builder/DockerignoreTest.java [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+    
+    @SuppressWarnings("resource")
+    @Test
+    public void testValidDockerignore() throws Exception {
+        ImageFromDockerfile img = new ImageFromDockerfile()
+                .withFileFromPath(".", DockerfileBuildTest.RESOURCE_PATH)
+                .withDockerfile(DockerfileBuildTest.RESOURCE_PATH.resolve("Dockerfile-currentdir"));
+        try (final GenericContainer<?> container = new GenericContainer<>(img.get())
+            .withStartupCheckStrategy(new OneShotStartupCheckStrategy())
+            .withCommand("ls", "/"))
```
Comment (bsideup) [inline core/src/test/java/org/testcontainers/images/builder/DockerignoreTest.java]: If would be nice to format it like here:
https://github.com/testcontainers/testcontainers-java/blob/a74b77f7dcd759f798d36f6b3b124f375c363bcf/core/src/test/java/org/testcontainers/junit/CopyFileToContainerTest.java#L15-L19

File: core/src/test/java/org/testcontainers/images/builder/DockerignoreTest.java [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+    @Test
+    public void testInvalidDockerignore() throws Exception {
+        try {
+            new ImageFromDockerfile()
+                .withFileFromPath(".", INVALID_DOCKERIGNORE_PATH)
+                .withDockerfile(INVALID_DOCKERIGNORE_PATH.resolve("Dockerfile"))
+                .get();
+            fail("Should not be able to build an image with an invalid .dockerignore file");
+        } 
+        catch(DockerClientException e) {
```
Comment (bsideup) [inline core/src/test/java/org/testcontainers/images/builder/DockerignoreTest.java]:
```suggestion
        catch (DockerClientException e) {
```

File: core/src/test/java/org/testcontainers/images/builder/DockerignoreTest.java (hunk 11 above)
Comment (bsideup) [reply core/src/test/java/org/testcontainers/images/builder/DockerignoreTest.java]: (maybe this a space between `try` and `(`

File: core/src/test/java/org/testcontainers/images/builder/DockerfileBuildTest.java [hunk 13]
```diff
@@ -56,12 +64,13 @@ public DockerfileBuildTest(String expectedFileContent, ImageFromDockerfile image
     public void performTest() {
         try (final GenericContainer container = new GenericContainer(image)
             .withStartupCheckStrategy(new OneShotStartupCheckStrategy())
-            .withCommand("cat", "/test.txt")) {
-
+            .withCommand("cat", "/test.txt")) 
+        {
```
Comment (bsideup) [inline core/src/test/java/org/testcontainers/images/builder/DockerfileBuildTest.java]: unrelated change?

File: core/src/main/java/org/testcontainers/images/builder/ImageFromDockerfile.java (hunk 10 above)
Comment (aguibert) [reply core/src/main/java/org/testcontainers/images/builder/ImageFromDockerfile.java]: good catch! My IDE (Eclipse) crossed out the method just by having `@deprecated` in the javadoc

File: core/src/test/java/org/testcontainers/images/builder/DockerignoreTest.java (hunk 12 above)
Comment (aguibert) [reply core/src/test/java/org/testcontainers/images/builder/DockerignoreTest.java]: I must say you are quite particular on code formatting @bsideup ;) Perhaps if you want to enforce this strict of formatting conventions, you could add a checkstyle plugin to the project?

File: core/src/test/java/org/testcontainers/images/builder/DockerfileBuildTest.java (hunk 13 above)
Comment (aguibert) [reply core/src/test/java/org/testcontainers/images/builder/DockerfileBuildTest.java]: yes, this is an unrelated change. Just cleaning up formatting to match what you had requested in another review comment since I was changing this file already

**Comment by aguibert** [comment]
rebased branch so it's up-to-date with master. Should be good to merge now

File: core/src/main/java/org/testcontainers/images/builder/ImageFromDockerfile.java [hunk 14]
```diff
@@ -142,7 +154,9 @@ public void onNext(BuildResponseItem item) {
 
     protected void configure(BuildImageCmd buildImageCmd) {
         buildImageCmd.withTag(this.getDockerImageName());
+        // always use withDockerfile because it honors .dockerignores and withDockerfilePath does not
```
Comment (kiview) [inline core/src/main/java/org/testcontainers/images/builder/ImageFromDockerfile.java]:
```suggestion
```
I think this comment isn't necessary here any more.
Does it make sense that we would set both `withDockerfilePath` and `withDockerfile` on `buildImageCmd` if both are set in `ImageFromDockerfile`?

Shouldn't one override the other?

Comment (aguibert) [reply core/src/main/java/org/testcontainers/images/builder/ImageFromDockerfile.java]: ultimately one will override the other, but we'll let the Docker Java API decide which one takes precedence in the event the user has set both of them. Ideally we could redirect the now-deprecated `withDockerfilePath` method to `withDockerfile` in our layer, but that may be a breaking change for some users so we can't do it

**Comment by rnorth** [comment]
That's 3 approvals, I'll merge. Thanks for another great contribution @aguibert, and sorry that it's taken a while to get this in.
